plugins/inputs/varnish/varnish.go

```diff
@@ -165,19 +272,187 @@ func (s *Varnish) Gather(acc telegraf.Accumulator) error {
+			} else {
+				metricValue = value
+			}
+		}
+
+		if parseError != nil {
+			acc.AddError(parseError)
+			continue
+		}
+
+		metric := s.parseMetricV2(fieldName)
+		if metric.vclName != "" && activeVcl != "" && metric.vclName != activeVcl {
+			//skip not active vcl
+			continue
+		}
 
+		fields := make(map[string]interface{})
+		fields[metric.fieldName] = metricValue
+		switch flag {
+		case "c", "a":
+			acc.AddCounter(metric.measurement, fields, metric.tags, timestamp)
+		case "g":
+			acc.AddGauge(metric.measurement, fields, metric.tags, timestamp)
+		default:
+			acc.AddGauge(metric.measurement, fields, metric.tags, timestamp)
+		}
+	}
 	return nil
 }
 
+// Parse the output of "varnishadm vcl.list -j" and find active vcls
+func getActiveVCLJson(out io.Reader) (string, error) {
+	var output = ""
+	if b, err := io.ReadAll(out); err == nil {
+		output = string(b)
+	}
+	// workaround for non valid json in varnish 6.6.1 https://github.com/varnishcache/varnish-cache/issues/3687
+	if strings.HasPrefix(output, "200") {
+		output = strings.TrimPrefix(output, "200")
+	}
+
+	var jsonOut []interface{}
+	err := json.Unmarshal([]byte(output), &jsonOut)
+	if err != nil {
+		return "", err
+	}
+
+	//check vcl.list output
+	if jsonOut[0].(float64) != 2 {
+		return "", fmt.Errorf("unsupported varnishadm format %v", jsonOut[0])
+	}
+
+	command := jsonOut[1].([]interface{})
+
+	if command[0] != "vcl.list" {
+		return "", fmt.Errorf("unsupported varnishadm command %v", jsonOut[1])
+	}
+	for _, s := range jsonOut {
+		switch s.(type) {
+		case map[string]interface{}:
+			vclStruct := s.(map[string]interface{})
+			if vclStruct["status"] == "active" {
+				return vclStruct["name"].(string), nil
+			}
+		default:
+			//ignore
+			continue
+		}
+	}
+	return "", nil
+}
+
+// Gets the "counters" section from varnishstat json (there is change in schema structure in varnish 6.5+)
+func getCountersJSON(rootJSON map[string]interface{}) map[string]interface{} {
+	//version 1 contains "counters" wrapper
+	if counters, exists := rootJSON["counters"]; exists {
+		return counters.(map[string]interface{})
+	}
+	return rootJSON
+}
+
+// converts varnish metrics name into field and list of tags
+func (s *Varnish) parseMetricV2(name string) (metric varnishMetric) {
+	metric.measurement = measurementNamespace
+	if strings.Count(name, ".") == 0 {
+		return metric
+	}
+	metric.fieldName = name[strings.LastIndex(name, ".")+1:]
+	var section = strings.Split(name, ".")[0]
+	metric.tags = map[string]string{
+		"section": section,
+	}
+
+	//parse name using regexpsCompiled
+	for _, re := range s.regexpsCompiled {
+		submatch := re.FindStringSubmatch(name)
+		if len(submatch) < 1 {
+			continue
+		}
+		for _, sub := range re.SubexpNames() {
+			if sub == "" {
+				continue
+			}
+			val := submatch[re.SubexpIndex(sub)]
+			if sub == "_vcl" {
+				metric.vclName = val
+			} else if sub == "_field" {
+				metric.fieldName = val
+			} else if val != "" {
+				metric.tags[sub] = val
+			}
```

Comment:
This is where `id` is getting added as a tag and why I am asking at a high level why we are increasing the cardinality on that value. Is that something users want to index against?

It seems concerning to add anything that doesn't match `_vcl` or `_field` as a tag as any change to the data format is going to cause cardinality to increase.

Reply:
Thanks for review. Id tag is added by the last default regexp rule (?P<id>)

```go
//generic metric like MSE_STORE.store-1-1.g_aio_running_bytes_write
regexp.MustCompile(`([\w\-]*)\.(?P<id>[\w\-.]*)\.([\w\-]*)`),
```

My idea was not to increase the cardinality, but make the field names corresponding with https://varnish-cache.org/docs/trunk/reference/varnish-counters.html

Cardinality issue was caused because varnish adds timestamps into its metrics names when reloading. This should be solved by tripping vcl name from the name. 

Example: we have following varnish metrics
```
  "MEMPOOL.busyobj.allocs"
  "MEMPOOL.ssl_buf.allocs"
  "MEMPOOL.req0.allocs"
  "MEMPOOL.sess0.allocs"
  "MEMPOOL.req1.allocs"
  "MEMPOOL.sess1.allocs"
```

Middle part identifies the name of resource. There are limited number of these so it is suitable for tagging/indexing.
We reduce number of tags but increase number of fields. I think cardinality in this case is the same. 

varnish,section=MEMPOOL,id=busyobj alloc=xxxx
varnish,section=MEMPOOL,id=ssl_buf alloc=xxxx
varnish,section=MEMPOOL,id=req0 alloc=xxxx

varnish,section=MEMPOOL busyobj.alloc=xxxx
varnish,section=MEMPOOL ssl_buf.alloc=xxxx
varnish,section=MEMPOOL req0.alloc=xxxx

Biggest disadvantage of this approach i see is that this approach will not be backward compatible with v1.
Varnish metrics can have various names, it tried to cover that in 'plugins/inputs/varnish/varnish_test.go:441' 

@helenosheaa @ismaelpuerto @anthosz  So should we remove the id tag and add middle part into field name?

Reply:
If there's a static number of values that will be set as id=tags I don't think it will be as much of cardinality problem as we were seeing with vcl name where fields included timestamps or ip's so they were constantly increasing. 

However what is the advantage of splitting it out into a tag? The more consistency between v1 and v2 unless there is a value add I think the better.

That said I believe the build from this PR is being tested so shall we make a decision on this once we get feedback from that?

Reply:
Splitting varnish metric names into tags can be useful when creating dashboards. Represent a newly added resource as a tag looks more natural to me, it will be automatically shown in the dashboard if you use generic queries and filters. 

But, may be, more important is to keep v2 as much as possible backward compatible and not break existing things. 
User can always use custom regexps in telegraf configuration to create custom tag splitting if needed.